### ISSUE TITLE
Incorrect behavior in tests with escaped filter chars

### ISSUE THREAD

**Issue by kiddick** [cannatag/ldap3#445]
Due to user able to use `( ) *` in DN of ldap entries, we need to escape in search filters.
So I before putting user query string in `search_filter` I am using  `from ldap3.utils.conv import escape_filter_chars`.
```
user_input = ')'
search_filter='(cn={})'.format(escape_filter_chars(user_input))
```
It work correctly with real connection. But in tests with `MOCK_ASYNC` strategy I am getting error
```
  File "ldap3-2.4-py3.5.egg/ldap3/core/connection.py", line 765, in search
    response = self.post_send_search(self.send('searchRequest', request, controls))
  File "ldap3-2.4-py3.5.egg/ldap3/strategy/mockAsync.py", line 111, in post_send_search
    responses, result = self.mock_search(request, controls)
  File "ldap3-2.4-py3.5.egg/ldap3/strategy/mockBase.py", line 612, in mock_search
    response, result =  self._execute_search(request)
  File "ldap3-2.4-py3.5.egg/ldap3/strategy/mockBase.py", line 652, in _execute_search
    filter_root = parse_filter(request['filter'], self.connection.server.schema, auto_escape=True, auto_encode=False)
  File "ldap3-2.4-py3.5.egg/ldap3/operation/search.py", line 220, in parse_filter
    raise LDAPInvalidFilterError('invalid filter')
ldap3.core.exceptions.LDAPInvalidFilterError: invalid filter
```

Seems there is an issue with parsing filter in mock connection. It fails on counting brackets, but it shouldn't.

**Comment by cannatag** [comment]
Please check code in dev. It should work now.

**Comment by cannatag** [comment]
Released in version 2.4.1
